code/modules/recycling/disposal/holder.dm

```diff
@@ -50,71 +57,91 @@
 				destinationTag = tagger.currTag
 
 
-// start the movement process
-// argument is the disposal unit the holder started in
+/// start the movement process.
+/// argument is the disposal unit the holder started in
 /obj/structure/disposalholder/proc/start(obj/machinery/disposal/D)
 	if(!D.trunk)
 		D.expel(src) // no trunk connected, so expel immediately
 		return
-	forceMove(D.trunk)
+	forceMove(D.trunk.associated_loc)
 	active = TRUE
 	setDir(DOWN)
-	move()
+	move(D.trunk)
 
-// movement process, persists while holder is moving through pipes
-/obj/structure/disposalholder/proc/move()
+/// movement process, persists while holder is moving through pipes
+/obj/structure/disposalholder/proc/move(obj/structure/disposalpipe/starting_node)
 	set waitfor = FALSE
 	var/ticks = 1
-	var/obj/structure/disposalpipe/last
+	var/obj/structure/disposalpipe/last_node = starting_node
 	while(active)
-		var/obj/structure/disposalpipe/curr = loc
-		last = curr
+		var/obj/structure/disposalpipe/current_node = last_node.transfer(src)
+
 		set_glide_size(DELAY_TO_GLIDE_SIZE(ticks * world.tick_lag))
-		curr = curr.transfer(src)
-		if(!curr && active)
-			last.expel(src, get_turf(src), dir)
 
-		ticks = stoplag()
+		if(!current_node && active)
+			last_node.expel(src, get_turf(src), dir)
+		last_node = current_node
+
+		sleep(1)
+		//ticks = stoplag()
 		if(!(count--))
 			active = FALSE
 
-//failsafe in the case the holder is somehow forcemoved somewhere that's not a disposal pipe. Otherwise the above loop breaks.
+//failsafe in the case the holder is somehow forcemoved somewhere that doesnt have a disposals pipe. Otherwise the above loop breaks.
 /obj/structure/disposalholder/Moved(atom/oldLoc, dir)
 	. = ..()
 	var/static/list/pipes_typecache = typecacheof(/obj/structure/disposalpipe)
 	//Moved to nullspace gang
-	if(!loc)
+
+	var/turf/turf_loc = get_turf(src)
+
+	if(!turf_loc)
+		return
+
+	var/has_possible_loc = FALSE
+
+	for(var/obj/structure/disposalpipe/possible_loc in turf_loc?.nullspaced_contents)
+		if(pipes_typecache[possible_loc.type])
```

Comment:
why not use the typecache var? also why does this check exist in the first place, it's already typecast

Reply:
lol